packages/strapi/lib/services/entity-validator/index.js

```diff
@@ -9,100 +9,132 @@ const _ = require('lodash');
 const { yup, formatYupErrors } = require('strapi-utils');
 const validators = require('./validators');
 
-module.exports = ({ strapi }) => ({
-  /**
-   * Validate some input based on a model schema
-   * @param {Object} model model schema
-   * @param {Object} data input data
-   * @param options
-   */
-  async validateEntity(model, data, options = { isDraft: false }) {
-    const validator = createValidator(model, options);
-
-    return validator
-      .validate(data, {
-        abortEarly: false,
-      })
-      .catch(error => {
-        throw strapi.errors.badRequest('ValidationError', { errors: formatYupErrors(error) });
-      });
-  },
-
-  /**
-   * Validate some input for updating based on a model schema
-   * @param {Object} model model schema
-   * @param {Object} data input data
-   * @param options
-   */
-  async validateEntityUpdate(model, data, options = { isDraft: false }) {
-    const validator = createUpdateValidator(model, options);
-
-    return validator
-      .validate(data, {
-        abortEarly: false,
-      })
-      .catch(error => {
-        throw strapi.errors.badRequest('ValidationError', { errors: formatYupErrors(error) });
-      });
-  },
-});
-
-const isMedia = attr => {
-  return (attr.collection || attr.model) === 'file' && attr.plugin === 'upload';
-};
+const isMedia = attr => (attr.collection || attr.model) === 'file' && attr.plugin === 'upload';
 
-const createValidator = (model, { isDraft }) => {
-  return yup
-    .object(
-      _.mapValues(model.attributes, attr => {
-        if (isMedia(attr)) {
-          return yup.mixed().nullable();
-        }
+const isSimpleAttribute = attr =>
+  !attr.collection && !attr.model && attr.type !== 'component' && attr.type !== 'dynamiczone';
 
-        let validator = createAttributeValidator(attr, { isDraft }).nullable();
+const addMinMax = (attr, validator, data) => {
+  if (Number.isInteger(attr.min) && (attr.required || (Array.isArray(data) && data.length > 0))) {
+    validator = validator.min(attr.min);
+  }
+  if (Number.isInteger(attr.max)) {
+    validator = validator.max(attr.max);
+  }
+  return validator;
+};
 
-        if (_.has(attr, 'default')) {
-          validator = validator.default(attr.default);
-        }
+const addRequiredValidation = createOrUpdate => (required, validator) => {
+  if (required) {
+    if (createOrUpdate === 'creation') {
+      validator = validator.notNil();
+    } else if (createOrUpdate === 'update') {
+      validator = validator.notNull();
+    }
+  } else {
+    validator = validator.nullable();
+  }
+  return validator;
+};
 
-        if (!isDraft && attr.required) {
-          return validator.notNil();
-        }
+const preventCast = validator => validator.transform((val, originalVal) => originalVal);
```

Comment:
`yup` casts some values if it can, it is useful to cast `1` to `true` when the expected type is `boolean`. However it does it also when we don't want to (when it's a dz, a component etc.), so I disabled it.
`val` is the casted value, `originalVal` is the original value.